File: src/helpers/helper2.js
const fs = require('fs');
const path = require('path');
const zmq = require('zeromq');
const uuidV4 = require('uuid/v4');
const child_process = require("child_process");
var initTracer = require('jaeger-client').initTracer;
var opentracing = require('opentracing');

// See schema https://github.com/jaegertracing/jaeger-client-node/blob/master/src/configuration.js#L37
var config = {
    serviceName: 'room-service',
    'reporter': {
        'logSpans': true,
        'agentHost': 'localhost',
        'agentPort': 6832
    },
    'sampler': {
        'type': 'const',
        'param': 1.0
    }
};
var options = {
    // metrics: metrics,
    // logger: logger,
};
// var tracer = initTracer(config, options);  // uncomment to use real tracer
var tracer = new opentracing.Tracer();  // no-op dummy tracer

const randomId = () => 
    uuidV4();

function sleep(millis) {
    return new Promise(resolve => setTimeout(resolve, millis));
}

const stringToTerm = x => {
    if (x[0] === `"`) {
        return ["text", x.slice(1, -1)]
    }
    if (isNaN(x) || x === "") {
        if (x[0] === "#") {
            return ["id", x.slice(1)]
        }
        if (x[0] === "$") {
            return ["variable", x.slice(1)]
        }
        if (x[0] === "%") {
            return ["postfix", x.slice(1)]
        }
        return ["text", x]
    }
    if (x.indexOf(".") === -1) {
        return ["integer", (+x).toString()]
    }
    return ["float", (+x).toFixed(6)]
}

const tokenizeString = str => {
    // from https://stackoverflow.com/questions/2817646/javascript-split-string-on-space-or-on-quotes-to-array
    var spacedStr = str.trim().replace(/\)/g, ' ) ').replace(/\(/g, ' ( ').replace(/,/g, ' , ').trim();
    var aStr = spacedStr.match(/[\w,()@#:\-\.\$\%]+|"[^"]+"/g), i = aStr.length;
    while(i--){
        aStr[i] = aStr[i].replace(/"/g,"");
    }
    return aStr
}

const fullyParseFact = q => {    
    if (typeof q === "string") {
        const q_tokens = tokenizeString(q)
        return q_tokens.map(x => stringToTerm(x))
    } else if (Array.isArray(q)) {
        let terms = [];
        for (var i = 0, len = q.length; i < len; i++) {
            if (typeof q[i] === "string") {
                const q_tokens = tokenizeString(q[i])
                terms = terms.concat(q_tokens.map(x => stringToTerm(x)))
            } else {
                terms = terms.concat([q[i]])
            }
        }
        return terms
    }
}

function getIdFromProcessName(scriptName) {
    return (scriptName.split(".")[0]).split("__")[0]
}

function getIdStringFromId(id) {
    return String(id).padStart(4, '0')
}

class Illumination {
    constructor() {
        this.illuminations = [];
        this.add = (type, opts) => {
            this.illuminations.push({ "type": type, "options": opts })
        }
        this.addColorType = (type, opts) => {
            opts = (opts.length === 1) ? opts[0] : opts;
            this.add(type, opts);
        }
    }
    rect(x, y, w, h) { this.add("rectangle", {"x": x, "y": y, "w": w, "h": h}) }
    ellipse(x, y, w, h) { this.add("ellipse", { "x": x, "y": y, "w": w, "h": h }) }
    text(x, y, txt) { this.add("text", { "x": x, "y": y, "text": txt }) }
    line(x1, y1, x2, y2) { this.add("line", [x1, y1, x2, y2]) }
    // point format: [[x1, y1], [x2, y2], ...]
    polygon(points) { this.add("polygon", points) }
    image(x, y, w, h, base64image) { this.add("image", { "x": x, "y": y, "w": w, "h": h, "bitmap_image_base64": base64image }) }
    // color format: string, [r, g, b], or [r, g, b, a]
    fill(...color) { this.addColorType("fill", color) }
    stroke(...color) { this.addColorType("stroke", color) }
    nostroke() { this.add("nostroke", []) }
    nofill() { this.add("nofill", []) }
    strokewidth(width) { this.add("strokewidth", width) }
    fontsize(width) { this.add("fontsize", width) }
    fontcolor(...color) { this.addColorType("fontcolor", color) }
    push() { this.add("push", []) }
    pop() { this.add("pop", []) }
    translate(x, y) { this.add("translate", { "x": x, "y": y }) }
    rotate(radians) { this.add("rotate", radians) }
    scale(x, y) { this.add("scale", { "x": x, "y": y }) }
    set_transform(m11, m12, m13, m21, m22, m23, m31, m32, m33) {
        this.add("transform", [m11, m12, m13, m21, m22, m23, m31, m32, m33])
    }
    toString() { return JSON.stringify(this.illuminations) }
}

function init(filename) {
    const scriptName = path.basename(filename);
    const scriptNameNoExtension = path.parse(scriptName).name;
    const logPath = filename.replace(scriptName, 'logs/' + scriptNameNoExtension + '.log')
    const access = fs.createWriteStream(logPath)
    process.stdout.write = process.stderr.write = access.write.bind(access);
    process.on('uncaughtException', function (err) {
        console.error((err && err.stack) ? err.stack : err);
    })
    const myId = getIdFromProcessName(scriptName);

    const rpc_url = process.env.PROG_SPACE_SERVER_URL || "localhost";
    const MY_ID_STR = getIdStringFromId(myId);
    client = zmq.socket('dealer');
    client.identity = MY_ID_STR;
    client.connect(`tcp://${rpc_url}:5570`);

    let init_ping_id = randomId()
    let select_ids = {}
    let subscription_ids = {}
    let server_listening = false
    let sent_ping = false;
    let batched_calls = []
    const DEFAULT_SUBSCRIPTION_ID = 0;
    let currentSubscriptionId = DEFAULT_SUBSCRIPTION_ID;
    var wireCtx;

    const sleep = (ms) => {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    const waitForServerListening = () => {
        // Sends a single ping and loops until server_listening is set to true
        // server_listening is set to true outside this function in the message received callback
        return new Promise(async resolve => {
            if (server_listening === false) {
                if (sent_ping == false) {
                    client.send([`.....PING${MY_ID_STR}${init_ping_id}`])
                    sent_ping = true;
                }
                while (server_listening === false) {
                    await sleep(100);
                    // await null; // prevents app from hanging
                }
                // console.log(`wait for server listening done ${MY_ID_STR}`)
            }
            resolve();
        });
    }

    const room = {
        setCtx: ctx => {
            wireCtx = ctx;
            this.wireCtx = ctx;
        },
        wireCtx: () => {
            return this.wireCtx;
        },
        onRaw: async (...args) => {
            await waitForServerListening();
            const query_strings = args.slice(0, -1)
            const callback = args[args.length - 1]
            const subscription_id = randomId()
            const query_msg = {
                "id": subscription_id,
                "facts": query_strings
            }
            const query_msg_str = JSON.stringify(query_msg)
            subscription_ids[subscription_id] = callback
            client.send([`SUBSCRIBE${MY_ID_STR}${query_msg_str}`]);
        },
        onGetSource: async (...args) => {
            const sourceVariableName = args[0]
            const query_strings = args.slice(1, -1).map(s => `$${sourceVariableName} $ ${s}`)
            const callback = args[args.length - 1]
            room.onRaw(...query_strings, callback)
        },
        on: async (...args) => {
            const query_strings = args.slice(0, -1).map(s => `$ $ ${s}`)
            const callback = args[args.length - 1]
            room.onRaw(...query_strings, callback)
        },
        select: async (...args) => {
            await waitForServerListening();
            const query_strings = args.slice(0, -1)
            const callback = args[args.length - 1]
            const select_id = randomId()
            const query_msg = {
                "id": select_id,
                "facts": query_strings
            }
            const query_msg_str = JSON.stringify(query_msg)
            select_ids[select_id] = callback
            client.send([`...SELECT${MY_ID_STR}${query_msg_str}`]);
        },
        assertNow: (fact) => {
            client.send([`....CLAIM${MY_ID_STR}${fact}`]);
        },
        assertForOtherSource: (otherSource, fact) => {
            batched_calls.push({
                "type": "claim",
                "fact": [
                    ["id", otherSource],
                    ["id", `${DEFAULT_SUBSCRIPTION_ID}`]
                ].concat(fullyParseFact(fact))
            })
        },
        assert: (...args) => {
            batched_calls.push({
                "type": "claim",
                "fact": [
                    ["id", MY_ID_STR],
                    ["id", `${currentSubscriptionId}`]
                ].concat(fullyParseFact(args))
            })
        },
        globalAssert: (...args) => {
            batched_calls.push({
                "type": "claim",
                "fact": [
                    ["id", MY_ID_STR],
                    ["id", `0`]
                ].concat(fullyParseFact(args))
            })
        },
        retractNow: (query) => {
            client.send([`..RETRACT${MY_ID_STR}${query}`]);
        },
        retractRaw: (...args) => {
            // TODO: need to push into an array specific to the subsciber, in case there are multiple subscribers in one client
            batched_calls.push({ "type": "retract", "fact": fullyParseFact(args) })
        },
        retractMine: (...args) => {
            if (typeof args === "string") {
                room.retractRaw(`#${MY_ID_STR} $ ${args}`)
            } else if (Array.isArray(args)) {
                room.retractRaw(...[["id", MY_ID_STR], ["variable", ""]].concat(args))
            }
        },
        retractMineFromThisSubscription: (...args) => {
            if (typeof args === "string") {
                room.retractRaw(`#${MY_ID_STR} #${currentSubscriptionId} ${args}`)
            } else if (Array.isArray(args)) {
                room.retractRaw(...[["id", MY_ID_STR], ["id", `${currentSubscriptionId}`]].concat(args))
            }
        },
        retractFromSource: (...args) => {
            const source = args[0]
            const retractArgs = args.slice(1);
            if (typeof retractArgs === "string") {
                room.retractRaw(`#${source} $ ${retractArgs}`)
            } else if (Array.isArray(retractArgs)) {
                room.retractRaw(...[["id", `${source}`], ["variable", ""]].concat(retractArgs))
            }
        },
        retractAll: (...args) => {
            if (typeof args === "string") {
                room.retractRaw(`$ $ ${args}`)
            } else if (Array.isArray(args)) {
                room.retractRaw(...[["variable", ""], ["variable", ""]].concat(args))
            }
        },
        flush: () => {
            // TODO: need to push into an array specific to the subsciber, in case there are multiple subscribers in one client
            room.batch(batched_calls);
            batched_calls = [];
        },
        batch: batched_calls => {
            const fact_str = JSON.stringify(batched_calls)
            client.send([`....BATCH${MY_ID_STR}${fact_str}`]);
        },
        cleanup: () => {
            room.retractMine(`%`)
        },
        cleanupOtherSource: (otherSource) => {
            const fact_str = JSON.stringify([{ "type": "death", "fact": [["id", otherSource]] }])
            client.send([`....BATCH${MY_ID_STR}${fact_str}`]);
        },
        draw: (illumination, target) => {
            target = typeof target === 'undefined' ? myId : target;
            room.assert(`draw graphics`, ["text", illumination.toString()], `on ${target}`)
        },
        newIllumination: () => {
            return new Illumination()
        },
        subscriptionPrefix: id => {
            currentSubscriptionId = id;
            room.retractMineFromThisSubscription(["postfix", ""])
        },
        subscriptionPostfix: () => {
            currentSubscriptionId = 0;
        }
    }

    const parseResult = result_str => {
        return JSON.parse(result_str).map(result => {
            let newResult = {}
            for (let key in result) {
                // result[key] = ["type", "value"]
                // for legacy reasons, we just want the value
                const value_type = result[key][0]
                if (value_type === "integer" || value_type === "float") {
                    newResult[key] = +result[key][1]    
                } else {
                    newResult[key] = result[key][1]
                }
            }
            return newResult
        })
    }

    client.on('message', (request) => {
        const span = tracer.startSpan(`client-${myId}-recv`, { childOf: room.wireCtx() });
        const msg = request.toString();
        // console.log(msg)
        const source_len = 4
        const SUBSCRIPTION_ID_LEN = (randomId()).length
        const SERVER_SEND_TIME_LEN = 13
        const id = msg.slice(source_len, source_len + SUBSCRIPTION_ID_LEN)
        const val = msg.slice(source_len + SUBSCRIPTION_ID_LEN + SERVER_SEND_TIME_LEN)
        if (id == init_ping_id) {
            server_listening = true
            console.log(`SERVER LISTENING!! ${MY_ID_STR} ${val}`)
            room.setCtx(tracer.extract(opentracing.FORMAT_TEXT_MAP, {"uber-trace-id": val}));
        } else if (id in select_ids) {
            const callback = select_ids[id]
            delete select_ids[id]
            callback(val)
        } else if (id in subscription_ids) {
            callback = subscription_ids[id]
            // room.cleanup()
            const r = parseResult(val)
            // const callbackSpan = tracer.startSpan(`client-${myId}-callbackrecv`, { childOf: span });
            callback(r)
            // callbackSpan.finish();
            room.flush()
        } else {
            console.log("unknown subscription ID...")
        }
        span.finish();
    });

    const run = async () => {
        await waitForServerListening();
        room.flush()
    }

    const afterServerConnects = async (callback) => {
        await waitForServerListening();
        callback();
    }

    return {
        room, myId, scriptName, MY_ID_STR, run, getIdFromProcessName, getIdStringFromId, tracer, afterServerConnects
    }
}

module.exports = init

File: src/helpers/helper2.py
import time
import zmq
import logging
import json
import uuid
import random
import os
import sys
import opentracing
from opentracing import Format, Tracer
from jaeger_client import Config



def init_jaeger_tracer():
    # config = Config(
    #     config={
    #         'sampler': {
    #             'type': 'const',
    #             'param': 1,
    #         },
    #         'logging': True,
    #     },  
    #     service_name='room-service',
    #     validate=True,
    # )
    # return config.initialize_tracer()
    return Tracer()  # no-op tracer

# this call also sets opentracing.tracer
tracer = None
ROOM_SPAN_CONTEXT = None

context = zmq.Context()
rpc_url = os.getenv('PROG_SPACE_SERVER_URL', "localhost")
client = context.socket(zmq.DEALER)

MY_ID = None
MY_ID_STR = None
SUBSCRIPTION_ID_LEN = len(str(uuid.uuid4()))
init_ping_id = str(uuid.uuid4())
server_listening = False
select_ids = {}
subscription_ids = {}

py_subscriptions = []
py_prehook = None

def get_my_id_str():
    global MY_ID_STR
    return MY_ID_STR

def get_my_id_pre_init(root_filename):
    global MY_ID, MY_ID_STR
    scriptName = os.path.basename(root_filename)
    scriptNameNoExtension = os.path.splitext(scriptName)[0]
    fileDir = os.path.dirname(os.path.realpath(root_filename))
    logPath = os.path.join(fileDir, 'logs/' + scriptNameNoExtension + '.log')
    logging.basicConfig(filename=logPath, level=logging.INFO)
    MY_ID = (scriptName.split(".")[0]).split("__")[0]
    return MY_ID


def override_my_id(id):
    global MY_ID, MY_ID_STR
    MY_ID = id
    MY_ID_STR = str(MY_ID).zfill(4)


def claim(fact_string):
    client.send_multipart(["....CLAIM{}{}".format(
        MY_ID_STR, fact_string)], zmq.NOBLOCK)


def batch(batch_claims):
    client.send_multipart(["....BATCH{}{}".format(
        MY_ID_STR, json.dumps(batch_claims)).encode()], zmq.NOBLOCK)


def retract(fact_string):
    client.send_multipart(["..RETRACT{}{}".format(
        MY_ID_STR, fact_string)], zmq.NOBLOCK)


def select(query_strings, callback):
    select_id = str(uuid.uuid4())
    query = {
        "id": select_id,
        "facts": query_strings
    }
    query_msg = json.dumps(query)
    select_ids[select_id] = callback
    msg = "...SELECT{}{}".format(MY_ID_STR, query_msg)
    logging.debug(msg)
    client.send_multipart([msg], zmq.NOBLOCK)


def subscribe(query_strings, callback):
    subscription_id = str(uuid.uuid4())
    query = {
        "id": subscription_id,
        "facts": query_strings
    }
    query_msg = json.dumps(query)
    subscription_ids[subscription_id] = callback
    msg = "SUBSCRIBE{}{}".format(MY_ID_STR, query_msg)
    logging.debug(msg)
    client.send_multipart([msg.encode()], zmq.NOBLOCK)


def parse_results(val):
    json_val = json.loads(val)
    results = []
    for result in json_val:
        new_result = {}
        for key in result:
            value_type = result[key][0]
            if value_type == "integer":
                new_result[key] = int(result[key][1])
            elif value_type == "float":
                new_result[key] = float(result[key][1])
            else:
                new_result[key] = str(result[key][1])
        results.append(new_result)
    return results


def listen(blocking=True):
    global server_listening, ROOM_SPAN_CONTEXT, tracer, MY_ID
    flags = 0
    if not blocking:
        flags = zmq.NOBLOCK
    try:
        raw_msg = client.recv_multipart(flags=flags)
    except zmq.Again:
        return False
    string = raw_msg[0].decode()
    span = tracer.start_span('client-'+MY_ID+'-recv', references=opentracing.child_of(ROOM_SPAN_CONTEXT))
    # preSpan = tracer.start_span('client-'+MY_ID+'-prerecv', child_of=span)
    source_len = 4
    server_send_time_len = 13
    id = string[source_len:(source_len + SUBSCRIPTION_ID_LEN)]
    val = string[(source_len + SUBSCRIPTION_ID_LEN +
                server_send_time_len):]
    # preSpan.finish()
    if id == init_ping_id:
        server_listening = True
        logging.info("SERVER IS LISTENING {}".format(MY_ID))
        # ROOM_SPAN_CONTEXT = tracer.extract(format=Format.TEXT_MAP, carrier={"uber-trace-id": val})
        # print(ROOM_SPAN_CONTEXT)
    elif id in select_ids:
        callback = select_ids[id]
        del select_ids[id]
        callback(val)
    elif id in subscription_ids:
        logging.debug(string)
        callback = subscription_ids[id]
        # callbackSpan = tracer.start_span('client-'+MY_ID+'-callbackrecv', child_of=span)
        callback(parse_results(val))
        # callbackSpan.finish()
    else:
        logging.info("UNRECOGNIZED:")
        logging.info(string)
    # span.finish()
    return True


def check_server_connection():
    global server_listening, client, init_ping_id, py_subscriptions, py_prehook
    SERVER_NO_RESPONSE_TIMEOUT_TIME_SECONDS = 2
    if server_listening:
        print("checking if server is still listening")
        server_listening = False
        init_ping_id = str(uuid.uuid4())
        listening_start_time = time.time()
        client.send_multipart([".....PING{}{}".format(MY_ID_STR, init_ping_id).encode()])
        while not server_listening:
            listen(blocking=False)
            # poll quickly went we think the server is already running and should respond
            time.sleep(0.01)
            if time.time() - listening_start_time > SERVER_NO_RESPONSE_TIMEOUT_TIME_SECONDS:
                # no response from server, assume server is dead
                check_server_connection()
                break
    else:
        print("SERVER DIED, attempting to reconnect")
        reconnect_check_delay_s = 10
        init_ping_id = str(uuid.uuid4())
        listening_start_time = time.time()
        client.send_multipart([".....PING{}{}".format(MY_ID_STR, init_ping_id).encode()])
        while not server_listening:
            print("checking if server is alive")
            listen(blocking=False)
            time.sleep(0.5)
            if time.time() - listening_start_time > SERVER_NO_RESPONSE_TIMEOUT_TIME_SECONDS:
                print("no response from server, sleeping for a bit...")
                time.sleep(reconnect_check_delay_s)
                client.send_multipart([".....PING{}{}".format(MY_ID_STR, init_ping_id).encode()])
        print("SERVER IS ALIVE!")
        if py_prehook:
            py_prehook()
        for s in py_subscriptions:
            query = s[0]
            callback = s[1]
            subscribe(query, callback)


def init(root_filename, skipListening=False):
    global MY_ID, MY_ID_STR, py_subscriptions, py_prehook, tracer
    scriptName = os.path.basename(root_filename)
    scriptNameNoExtension = os.path.splitext(scriptName)[0]
    fileDir = os.path.dirname(os.path.realpath(root_filename))
    logPath = os.path.join(fileDir, 'logs/' + scriptNameNoExtension + '.log')
    logging.basicConfig(filename=logPath, level=logging.INFO)
    if MY_ID_STR is None:
        MY_ID = (scriptName.split(".")[0]).split("__")[0]
        MY_ID_STR = str(MY_ID).zfill(4)
    print("INSIDE INIT:")
    print(MY_ID)
    print(MY_ID_STR)
    print("tcp://{0}:5570".format(rpc_url))
    # print(logPath)
    # print("-")
    tracer = init_jaeger_tracer()
    client.setsockopt(zmq.IDENTITY, MY_ID_STR.encode())
    client.connect("tcp://{0}:5570".format(rpc_url))
    print("connected")
    client.send_multipart([".....PING{}{}".format(MY_ID_STR, init_ping_id).encode()])
    print("sent ping")
    listen()  # assumes the first message recv'd will be the PING response

    # time.sleep(1.0)
    
    if py_prehook:
        py_prehook()
    # for s in selects:
    #     query = s[0]
    #     callback = s[1]
    #     select(query, callback)
    for s in py_subscriptions:
        query = s[0]
        callback = s[1]
        subscribe(query, callback)
    if skipListening:
        return
    while True:
        listen()


def prehook(func):
    global py_prehook
    py_prehook = func

    def function_wrapper(x):
        func(x)
    return function_wrapper


def subscription(expr):
    def subscription_decorator(func):
        global py_subscriptions
        py_subscriptions.append((expr, func))

        def function_wrapper(x):
            func(x)
        return function_wrapper
    return subscription_decorator


def tracer_cleanup():
    global tracer
    if tracer is not None:
        time.sleep(2)   # yield to IOLoop to flush the spans - https://github.com/jaegertracing/jaeger-client-python/issues/50
        tracer.close()  # flush any buffered spans

import atexit
atexit.register(tracer_cleanup)